### DISCUSSION TITLE
How can we add javascript at the end of the body?

### DISCUSSION THREAD

**Discussion by joshbedo** [vercel/next.js#14863] [Help]
I have a project where I need scripts to load at the end of the body to create carousels and other things. For some reason i can't get it to run correctly. It runs fine outside of Next.js. Is there something I'm missing?

**Comment by sauhard98** [comment]
```JS
const [scriptLoaded, setScriptLoaded] = React.useState(false);
//This is not necessary but a check if script is required in some other part of the code`
if (typeof window === undefined && !scriptLoaded) {
        const script = document.createElement('script');
        script.setAttribute('async', ''); // Or defer or nothing
        script.setAttribute('id', "script-id");
        script.src = "source-of-script";
        const position = document.querySelector("body"); // Or any other location , example head
        position.appendChild(script);
        setScriptLoaded(true);
}
```

**Reply by joshbedo** [reply]
This looks really hacky to me, there isn't another way?

**Reply by sauhard98** [reply]
I have used it myself and it works fine.
I found this in the Material-UI's official example([link](https://material-ui.com/components/autocomplete/#GoogleMaps.js)) and I trust google to keep it bug free.
They used useRef instead of useState hook.

**Reply by yamankatby** [reply]
Try this hook. Do the same
```js
export const useLoadScripts = (scripts) => {
  const scriptsLoaded = useRef(false);
  if (typeof window !== "undefined" && !scriptsLoaded.current) {
    scripts.forEach((script) => {
      const element = document.createElement("script");
      element.src = script;
      element.type = "text/javascript";
      const position = document.querySelector("body");
      position.appendChild(element);
    });
    scriptsLoaded.current = true;
  }
};
```

**Reply by lxsmnsyc** [reply]
@yamankatby iirc that is not Concurrent Mode-safe

**Reply by drksbr** [reply]
Nice

**Comment by cristiand391** [comment]
Next.js provides a built-in solution for these kind of use cases.

You can use a custom `Document` to add a script at the end of the body, just like you would do in html:

```
import Document, { Html, Head, Main, NextScript } from 'next/document'

class MyDocument extends Document {
  render() { 
    return (
      <Html>
	<Head />
	<body>
	  <Main />
	  <NextScript />
          <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" crossorigin="anonymous"></script>
	</body>
      </Html>
    )
  }
}

export default MyDocument;
```

[Next.js docs](https://nextjs.org/docs/advanced-features/custom-document)

**Reply by a2br** [reply]
You forgot to camelCase `crossOrigin` :)

**Reply by robert-chiniquy** [reply]
What if you only want this to occur in a single page, not every page? Kind of like adding to `<Head>` inside a page handler?

**Comment by nandorojo** [comment]
What is the best practice for this in Next.js? I see no answers on the docs, and the forums have all kind of answers on what the best way to import external scripts is.

**Reply by lxsmnsyc** [reply]
There's an upcoming `<Script />` component in Next that allows injecting scripts at the end of the body

**Reply by nring** [reply]
> There's an upcoming `<Script />` component in Next that allows injecting scripts at the end of the body

Sounds promising! Do you have a link to a discussion or PR?

**Reply by lxsmnsyc** [reply]
@nring I can't seem to find the actual RFC about this component, but it was introduced "partially" during the conference

Edit:
https://github.com/vercel/next.js/discussions/18172

**Comment by MineiToshio** [comment]
Starting next.js 11, you can use the Script Component https://nextjs.org/docs/basic-features/script.

**Reply by cristiand391** [reply]
Sweet, haven't used next.js in a while, will take a look at it.
Thanks!

**Comment by itsrehanraihan** [comment]
You better use Next.js ' Script ' component,
It is an extension of the HTML <script> element.
